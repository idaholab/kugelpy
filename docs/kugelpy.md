Copyright 2024, Battelle Energy Alliance, LLC, ALL RIGHTS RESERVED
# `kugelpy`

`kugelpy` is a python module that performs time-dependent high-temperature gas-cooled pebble bed reactor (PBR) analysis. 
This process can simulate the run-in procedure for a PBR along with equilibrium analysis. 
The following notebook will describe how to create a `kugelpy` input and how to load previous `kugelpy` simulations to restart or examine the state of the reactor.

The examples are run using a pbs script, see an example [here](./kugelpy/kugelpy.pbs), feel free to use this script and update it for your work. 
Typically the script should be run on a high performance computing cluster (HPC), as the simulations are computationally expensive (reasonable results have been shown between 30-40 nodes with 48 processors each).
The python script has been annotated to provide clarity for the input script.

## Examples

[Multi-Pass Jump-In Simulation](./jump_in.md)

[Multi-Pass Run-In Simulation](./run_in.md)

[OTTO Jump-In Simulation](./otto.md)

[Multiphysics Run-In Simulation](./mp_run_in.md)

[Loading a Previous Simulation](./loader.md)

### Variables

#### Questions

If you have questions or need help with the `kugelpy` module, please direct questions to Ryan Stewart (ryan.stewart@inl.gov).