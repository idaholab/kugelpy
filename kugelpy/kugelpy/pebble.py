'''
@authors: balep, stewryan, ethan-fowler

Copyright 2024, Battelle Energy Alliance, LLC, ALL RIGHTS RESERVED
'''

import numpy as np
import math

class Pebble(object):
    """
    Class for holding information for an individual pebble.
    
    Parameters
    ----------
    x: float
        current x position in the pebble bed
    y: float
        current y position in the pebble bed
    z: float
        current z position in the pebble bed
    r: float
        current radial position in the pebble bed
    radius: float
        Outer radius of the pebble surrounding the graphite matrix [cm]
    volume: int
        Integer representation the volume region the pebble is in 
    channel: int
        Integer representation the channel the pebble is in     
    
    Attributes
    ----------
    pebble_type: str
        Designation if this is a graphite or fuel pebble
    inner_radius: float
        Inner radius of the pebble surrounding the graphite matrix [cm]
    material: dict
        Dictionary of the material in the pebble. Takes the form:
        {'material_name': {'ZAID1': float, 'ZAID2': float, ...}}
    geometry: dict
        Dictionary of the geometry in the pebble. 
        Each entry describes a part of the geometry, where a given radius and number of instances are used to calculate the volume
        Takes the form:
        {'geometry_name': {'radius': float, 'number_of_instances': int, 'volume': float}}
    x: float
        current x position in the pebble bed
    y: float
        current y position in the pebble bed
    z: float
        current z position in the pebble bed
    r: float
        current radial position in the pebble bed
    radius: float
        Outer radius of the pebble surrounding the graphite matrix [cm]
    volume_num: int
        Integer representation the volume region the pebble is in 
    channel_num: int
        Integer representation the channel the pebble is in 
    mesh_location: str
        Combination of channel, and volume for where the pebble is
    xs_library: str
        Cross-section library for Serpent to use 
    scattering_library: str
        Name of the scattering library for Serpent to use
    temperature: float
        Temperature of the pebble [K]
    shuffled: Bool
        Indicator if this is the first pass or Nth pass through the core
    previous_universe: str
        Concatinated string containing the previous universes the pebble was exposed to
        If the pebble is homogenized after each pass, we default the previous universes to c0v0
    burnup: float
        Burnup accrued during pebble movement [MWd/kg]
    num_passes: int
        number of passes a pebble has gone through the core
    pebble_number: int
        integer related to the pebble number in the core
    universe: str
        string containing two integers, channel and volume
    """

    def __init__(self, x, y, z, r, radius, channel, volume, temperature=900.0, pass_limit=6, pebble_number=0):
        self._pebble_type = 'graphite'
        self._inner_radius = 2.5
        self._radius = radius
        self._material = {'matrix':   {'6000': 9.22570E-2, '5010': 2.28336E-9, '5011': 9.24876E-9},
                         'pebshell': {'6000': 6.61814E-2, '5010': 2.23273E-8,  '5011': 9.04368E-7}}
#        self._material = {'matrix':   {'6000': 9.22570E-2, },
#                         'pebshell': {'6000': 6.61814E-2, }}        
        self._x = x
        self._y = y
        self._z = z
        self._r = r
        self._volume_num = volume
        self._channel_num = channel
        self._mesh_location = f'c{channel}v{volume}'
        self._temperature = temperature
        self._shuffled = False
        self._previous_universe = ''
        self._burnup = 0.
        self._burnup_j_cm3 = 0.
#        self._burnup_group = 0
        self._num_passes = 0
        self._pass_limit = pass_limit
        self._pebble_number = pebble_number
        self._xs_dict = {1500: {'graphite': 'grph1500', 'xs_set': '15c'},
                        1200: {'graphite': 'grph1200', 'xs_set': '12c'},
                         900: {'graphite':  'grph900', 'xs_set': '09c'},
                         600: {'graphite':  'grph600', 'xs_set': '06c'},
                         300: {'graphite':  'grph300', 'xs_set': '03c'},}
        self._previous_universe = 'g0_' + self._mesh_location
        self.set_universe()
        self._xs_library, self._scattering_library = self.set_xs_set(self._temperature)

        self._geometry = {'matrix':    {'radius': self._inner_radius,    'number_of_instance': 1, 'volume': 0.0},
                         'pebshell':  {'radius': radius, 'number_of_instance': 1, 'volume': 0.0}}
        self.calculate_volumes()

    def calculate_volumes(self):
        """
        Calculate the volume for each constituent part of the pebble
        """
        previous_volume = 0.0
        for data in self._geometry.values():
            volume = math.pi * 4 / 3 * math.pow(data['radius'], 3) 
            total_region_volume = volume * data['number_of_instance']
            data['volume'] = total_region_volume - previous_volume
            previous_volume = total_region_volume 

    def set_xs_set(self, temperature):
        """
        Grab the correct cross-section set and scattering library based on the temperature
        """
        dict_array = np.array([x for x in self._xs_dict.keys()])
        temp = dict_array[dict_array<temperature].max() if temperature not in self._xs_dict.keys() else temperature
        return self._xs_dict[temp]['xs_set'], self._xs_dict[temp]['graphite']
        
    def update_position(self, x, y, z, r, channel, volume, shuffled=False):
        """
        Update the position of the pebbles after a shift or refuel.
        """
        self._shuffled = shuffled
        self._x = x
        self._y = y
        self._z = z
        self._r = r
        self._channel_num = channel
        self._volume_num = volume
        self._mesh_location = f'c{channel}v{volume}'
        self.set_universe()

    def set_previous_universe(self, prev_universe):
        """
        Update the previous universe after a shift or refuel
        """
        self._previous_universe = prev_universe

    def increase_pass(self):
        """
        When pebble is refueled, increse the number of passes it has gone through.
        """
        self._num_passes += 1
    
    def set_universe(self):
        """
        Update the current universe of the pebbble, if this is not the first pass (i.e. it has been shuffled, add the previous universe to the current pebble location
        """
        if self._shuffled:
            self._universe = self._previous_universe + '_' + self._mesh_location
        else:
            split_univ = self._previous_universe.split('_')[1:]
            univ = '' if len(split_univ) == 1 else '_'.join([x for x in split_univ[:-1]]) + '_'
            self._universe = f'g_' + univ + self._mesh_location 
            #self._universe = f'g_' + f'p{self._num_passes}_' + f'bu{self._burnup_group}_ + univ + self._mesh_location
    
    def update_pebble_temperature(self,pebble_temp,fuel_temp=None):
        """
        Update the pebble temperature and corresponding cross-section/scattering libraries.
        """
        self._temperature = pebble_temp
        self._xs_library, self._scattering_library = self.set_xs_set(pebble_temp)

class FuelPebble(Pebble):
    """
    Class for holding information for an individual pebble.
    
    Parameters
    ----------
    x: float
        current x position in the pebble bed
    y: float
        current y position in the pebble bed
    z: float
        current z position in the pebble bed
    r: float
        current radial position in the pebble bed
    radius: float
        Outer radius of the pebble surrounding the graphite matrix [cm]
    volume: int
        Integer represention the volume region the pebble is in 
    channel: int
        Integer represention the channel the pebble is in 
    fuel_material: dict
        Dictionary containing the fuel material, will replace 'fuel' in self._material
    
    
    Attributes
    ----------
    kernels_per_pebble: int
        Number of TRISO kernels per fuel pebble
    triso_volume: float
        Volume of the fuel kernels in the pebble (required for correct burnup)
    power_days: float
        Number of megawatt days the pebbles has been in the core (days * power) [MWd]
    power_density: float
        Current power density for the pebble (this is the power density for all pebbles of the same fuel type, same pass number, and in the same core location) [MW]
    geometry: dict

    """
    
    def __init__(self, x, y, z, r, radius, channel, volume, fuel_material=None, homogenization_group=0, temperature=900.0,  fuel_temperature=900.0, pass_limit=6, pebble_number=0, kernel_data=None):
        self._homogenization_group = homogenization_group
        super().__init__(x, y, z, r, radius, channel, volume, temperature=temperature, pass_limit=pass_limit, pebble_number=pebble_number)
        self._pebble_type = 'fuel'
        self._kernel_data = kernel_data if kernel_data else {'fuel':0.02125, 'buffer':0.03125, 'inner_pyc':0.03525, 'sic':0.03875, 'outer_pyc':0.04275, 'kernels_per_pebble': 18775}
        self._kernels_per_pebble = self._kernel_data.pop('kernels_per_pebble')
        self._material = {'fuel':      {'92235': 3.70100E-3, '92238': 1.99216E-2,  '8016': 3.37331E-1, '6000': 9.26006E-3, '5010': 1.87091E-8, '5011': 7.57813E-8},
                         'buffer':    { '6000': 5.26466E-2,  '5010': 1.35512E-8,  '5011': 5.48894E-8},
                         'inner_pyc': { '6000': 9.52653E-2,  '5010': 2.45213E-8,  '5011': 9.93236E-8},
                         'sic':       {'14028': 4.43271E-2, '14029': 2.25082E-3, '14030': 1.48376E-3, '6000': 4.80616E-2,  '5010': 1.23711E-8,  '5011': 5.0109E-8},
                         'outer_pyc': { '6000': 9.52653E-2,  '5010': 2.45213E-8,  '5011': 9.93236E-8},
                         'matrix':    { '6000': 8.67416E-2,  '5010': 2.23273E-8,  '5011': 9.04368E-8},
                         'pebshell':  { '6000': 8.67416E-2,  '5010': 2.23273E-8,  '5011': 9.04368E-8}}

        #self._material = {'fuel':      {'92235': 3.70100E-3, '92238': 1.99216E-2,  '8016': 3.37331E-1, '6000': 9.26006E-3, },
        #                 'buffer':    { '6000': 5.26466E-2, },
        #                 'inner_pyc': { '6000': 9.52653E-2,  },
        #                 'sic':       {'14028': 4.43271E-2, '14029': 2.25082E-3, '14030': 1.48376E-3, '6000': 4.80616E-2, },
        #                 'outer_pyc': { '6000': 9.52653E-2, },
        #                 'matrix':    { '6000': 8.67416E-2, },
        #                 'pebshell':  { '6000': 8.67416E-2, }}                         
        self._fuel_temperature = fuel_temperature
        self._previous_universe = f'f{self._homogenization_group}_' + self._mesh_location

        self.set_universe()
        self._previous_universe = self._universe
        self._triso_volume = self._kernels_per_pebble * 0.0000402
        self._power_days = 0.
        self._power_density = 0.
        self._days_in_core = 0
        if fuel_material:
            self._material['fuel'] = fuel_material
        self._xs_fuel_library, self._fuel_scattering_library = self.set_xs_set(self._fuel_temperature)

        self._geometry = {'fuel':      {'radius': self._kernel_data['fuel'],      'number_of_instance': self._kernels_per_pebble, 'volume': 0},
                         'buffer':    {'radius': self._kernel_data['buffer'],    'number_of_instance': self._kernels_per_pebble, 'volume': 0},
                         'inner_pyc': {'radius': self._kernel_data['inner_pyc'], 'number_of_instance': self._kernels_per_pebble, 'volume': 0},
                         'sic':       {'radius': self._kernel_data['sic'],       'number_of_instance': self._kernels_per_pebble, 'volume': 0},
                         'outer_pyc': {'radius': self._kernel_data['outer_pyc'], 'number_of_instance': self._kernels_per_pebble, 'volume': 0},
                         'matrix':    {'radius': self._inner_radius, 'number_of_instance': 1},
                         'pebshell':  {'radius': radius, 'number_of_instance': 1}}
        self.calculate_volumes()
        

    def update_fuel_material(self, material):
        """
        Update the fuel material
        """
        self._material['fuel'] = material

    def update_burnup(self, power, days):
        """
        Add up the total power days the pebble has been in the core.
        Divide by the kg of HM originally in the pebble (currently assumed to be 7 g or 0.007 kg)
        To obtain the burnup in J/cm^3 
        """
        self._days_in_core += days
        self._power_density = power / (1E6) # Convert from W to MW, power of an individual pebble
        self._power_days += (self._power_density * days)
        self._burnup =  self._power_days / 0.007 # MWd/kg
        self._burnup_j_cm3 = self._power_days / self._triso_volume * 1E6 * 86400# for J/cm^3 we divide the power days (Watt-day or J/s * days) by the kernel volume (cm^3) and convert to MJ from J (1E6) and from days to seconds (86,400 s/day)
        
    def set_universe(self):
        """
        Update the current universe of the pebbble, if this is not the first pass (i.e. it has been shuffled, add the previous universe to the current pebble location        
        """
        if self._shuffled:
            self._universe = self._previous_universe + '_' + self._mesh_location
        else:
            split_univ = self._previous_universe.split('_')[1:]
            univ = '' if len(split_univ) == 1 else '_'.join([x for x in split_univ[:-1]]) + '_'
            self._universe = f'f{self._homogenization_group}_' + univ + self._mesh_location   
            
    def update_pebble_temperature(self,pebble_temp,fuel_temp=None):
        """
        Update the temperatures (and corresponding cross-sections) for both the pebble and fuel materials.
        """
        self._temperature = pebble_temp
        self._fuel_temperature = fuel_temp
        self._xs_fuel_library, self._fuel_scattering_library = self.set_xs_set(fuel_temp)
        self._xs_library, self._scattering_library = self.set_xs_set(pebble_temp)
