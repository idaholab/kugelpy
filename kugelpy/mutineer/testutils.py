'''Utility module for testing modules.

Created on 2023-05-11 16:13:01

@authors: balep, stewryan, ethan-fowler

Copyright 2024, Battelle Energy Alliance, LLC, ALL RIGHTS RESERVED
'''
#===============================================================================
# Import
#===============================================================================
from _ctypes import ArgumentError
from os import path, mkdir,getenv
from errno import EEXIST

#===============================================================================
# Definitions
#===============================================================================
def gen_tmp_folder(main_tests_dir):
    if not path.isdir(main_tests_dir):
        try:
            mkdir(main_tests_dir)
        except OSError as exc:
            if exc.errno != EEXIST:
                raise
            pass

def find_pbed_input(file, file_path):
    '''
    Attempts to locate the start, end, and format of Serpent pbed position input in file.

    file: (str) name of file with pbed data 
    file_path: (str) path to file
    '''

    found_start = False # keeps track of whether or not start index had been found
    start_line = None # stores fist pbed geometry input line index
    end_line = None # stores last pbed geometry input line index
    format = None # indicates format of pbed geometry input line

    input = open(path.join(file_path, file)).readlines()
    
    for i in range(len(input)):
        temp = input[i].split()
        if len(temp)==6:
            try:
                float(temp[5])
            except:
                if not found_start:
                    found_start = True
                    format = 1
                    start_line = i
        elif len(temp)==5:
            try:
                float(temp[4])
            except:
                if not found_start:
                    found_start = True
                    start_line = i
                    format = 0
        else:
            if found_start:
                end_line = i
    
    return start_line, end_line, format

def compare_pbeds(solution_file, solution_path, test_file, test_path, solution_start_line=0, solution_end_line=None, format=0):
    '''
    Reads in the position (x, y, z) and radius of pebbles for comparison with expectation.
    In some edge cases, variable rounding due to numpy version can result in pebbles being placed in different 
    universes, thus universe names are not accounted for.

    solution_file: (str) file name of pre-generated solution output used for comparison
    solution_path: (str) path to solution output file
    test_file: (str) file name of pre-generated test output used for comparison
    test_path: (str) path to test output file
    solution_start_line: (str) index of line in solution file where pebble positions begin (0 by default)
    solution_end_line: (str) index of line in solution file where pebble positions end (None by default)
    format: (int) 1 if lines are numbered, 0 otherwise
    '''

    # read in solution and test files
    solution = open(path.join(solution_path, solution_file)).readlines()[solution_start_line:solution_end_line]
    test = open(path.join(test_path, test_file)).readlines()
    
    # remove new line characters
    while '\n' in solution:
        solution.remove('\n')
    while '\n' in test:
        test.remove('\n') 

    if len(solution) != len(test):
        # if files contain different number of pebbles, write failed
        print("compare_pbeds() Error: Different number of pebbles")
        print("Solution Length: ", len(solution))
        print("Test Length: ", len(test))
        return False
    else:
        # format solution and test lists
        for i in range(len(solution)):
            temp = solution[i].split()
            solution[i] = f'{temp[0+format]} {temp[1+format]} {temp[2+format]} {temp[3+format]}'
            temp = test[i].split()
            test[i] = f'{temp[0+format]} {temp[1+format]} {temp[2+format]} {temp[3+format]}'
        
        for line in test:
            # check to make sure every pebble in test is also in solution
            if line in solution:
                # remove line from solution to reduce number of comparison operations
                solution.remove(line)
            else:
                # pebble in line not contained in solution, write failed
                print("Incorrect line: ", line)
                print("Remaining length: ", len(solution))
                return False
        
        return True

#===============================================================================
# Main
#===============================================================================
if __name__ == '__main__':
    pass

