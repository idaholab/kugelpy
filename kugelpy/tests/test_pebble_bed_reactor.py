'''
@author: balep, stewryan, ethan-fowler

Copyright 2024, Battelle Energy Alliance, LLC, ALL RIGHTS RESERVED
'''

import kugelpy.kugelpy.kugelpy.pebble_bed_reactor as pbr
from os import remove, path
from kugelpy.kugelpy.mutineer.testutils import gen_tmp_folder

#===============================================================================
# Shared objects
#===============================================================================
# Paths.
main_dir = path.dirname(path.realpath(__file__))
ifiles_dir = path.join(main_dir, 'testfiles')
distfile_path = path.join(ifiles_dir, 'dist.txt')
main_tests_dir = path.join(main_dir, '_tmp/')
gen_tmp_folder(main_tests_dir)

def test_init(regtest):
    pbr_instance = pbr.PebbleBedReactor(cavity_height=100)
    print(vars(pbr_instance))
    print(vars(pbr_instance), file=regtest)

def test_build_block(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    pbr_instance.build_block(1, 0.0)
    print(pbr_instance._block_dict, file=regtest)

def test_build_all_blocks(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    pbr_instance.build_all_blocks()
    print(pbr_instance._block_dict, file=regtest)

def test_build_pbr_core(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    pbr_instance.build_pbr_core()
    print(pbr_instance._block_dict, file=regtest)
    print(pbr_instance._reactor_dict, file=regtest)

def test_build_cylinder_surface(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    cylv = pbr_instance.build_cylinder_surface('surf_1', 'cylv', 40, u=1, v=1, w=1)
    print(cylv, file=regtest)
    cylz = pbr_instance.build_cylinder_surface('surf_1', 'cylz', 40, lower_height=10, upper_height=2)
    print(cylz, file=regtest)
    cyly = pbr_instance.build_cylinder_surface('surf_1', 'cyly', 40, x_offset=10, y_offset=2, z_offset=5)
    print(cyly, file=regtest)
    cylx = pbr_instance.build_cylinder_surface('surf_1', 'cylx', 40, x_offset=10, y_offset=2, z_offset=5)
    print(cylx, file=regtest)

def test_build_conus(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    pbr_instance.build_conus()
    print(pbr_instance._reactor_dict, file=regtest)

def test_build_outlet_plenum(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    pbr_instance.build_outlet_plenum()
    print(pbr_instance._reactor_dict, file=regtest)

def test_build_outlet_channel(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    pbr_instance.build_outlet_channel()
    print(pbr_instance._reactor_dict, file=regtest)

def test_build_pebble_shoot(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    pbr_instance.build_pebble_shoot()
    print(pbr_instance._reactor_dict, file=regtest)

def test_build_bottom_reflector(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    pbr_instance.build_bottom_reflector()
    print(pbr_instance._reactor_dict, file=regtest)

def test_build_pebble_bed(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    pbr_instance.build_pebble_bed()
    print(pbr_instance._reactor_dict, file=regtest)

def test_build_pebble(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    pbr_instance.build_pebble()
    print(pbr_instance._reactor_dict, file=regtest)

def test_build_outside(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    pbr_instance.build_outside()
    print(pbr_instance._reactor_dict, file=regtest)

def test_build_cavity(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    pbr_instance.build_cavity()
    print(pbr_instance._reactor_dict, file=regtest)

def test_build_top_reflector(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    pbr_instance.build_top_reflector()
    print(pbr_instance._reactor_dict, file=regtest)

def test_build_cell(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    cyl = pbr_instance.build_cell('cell1', 'univ1', 'material', ['surf_0', 'surf_2'], outside_surfaces=['surf_5', 'surf_6'], outside_cells=['cell_0', 'cell_2'] )
    print(cyl, file=regtest)

def test_build_filled_cell(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    cyl = pbr_instance.build_filled_cell('cell1', 'univ1', 'material', ['surf_0', 'surf_2'], outside_surfaces=['surf_5', 'surf_6'], outside_cells=['cell_0', 'cell_2'] )
    print(cyl, file=regtest)

def test_build_universe(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    cyl = pbr_instance.build_universe('cellu1',  inside_surfaces=['surf_0', 'surf_2'], outside_surfaces=['surf_5', 'surf_6'], outside_cells=['cell_0', 'cell_2'] )
    print(cyl, file=regtest)

def test_convert_theta_to_uv(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    theta = pbr_instance.convert_theta_to_uv(0.0, r=50)
    print(theta, file=regtest)

def test_build_block_surfaces(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    surf = pbr_instance.build_block_surface(1, 2, 4, 0, 30, x_offset=1.0, y_offset=12.0)
    print(surf, file=regtest)

def test_build_dimple_surfaces(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    dimple = pbr_instance.build_dimple_surfaces(1, 3, 0, 15, x_offset= 1.0, y_offset=-10.0)
    print(dimple, file=regtest)

def test_build_rod_surfaces(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    rod = pbr_instance.build_rod_surfaces(10, 10, 3, 10, 'cr', 0.0, 100.0)
    print(rod, file=regtest)

def test_build_block_cell(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    cell = pbr_instance.build_block_cell(1, surfaces_to_skip=['surf1', 'surf2'], cells_to_skip=['cell1', 'cell2'])
    print(cell, file=regtest)

def test_build_dimple_cells(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    dimple = pbr_instance.build_dimple_cells('test', 1)
    print(dimple, file=regtest)

def test_build_rod_cells(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    rod_cells =  pbr_instance.build_rod_cells('test', 'cr', 'control', surfaces_to_skip=['sruf1', 'surf2'], cells_to_skip=['cell1', 'cell2'])
    print(rod_cells, file=regtest)

def test_block_universe(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    rod_cells =  pbr_instance.build_block_universe('test', surfaces_to_skip=['sruf1', 'surf2'], cells_to_skip=['cell1', 'cell2'])
    print(rod_cells, file=regtest)

def test_build_dimple_universes(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    dimple_cells =  pbr_instance.build_dimple_universes('test', 1)
    print(dimple_cells, file=regtest)

def test_build_rod_universe(regtest):
    pbr_instance = pbr.PebbleBedReactor()
    rod_univ =  pbr_instance.build_rod_universe('test', 'cr', surfaces_to_skip=['sruf1', 'surf2'], cells_to_skip=['cell1', 'cell2'])
    print(rod_univ, file=regtest)

def test_build_pbr_core(regtest):
    pbr_instance = pbr.PebbleBedReactor(output_dir=main_tests_dir)
    pbr_instance.build_pbr_core()
    pbr_instance.write_pbr_core()
    file_ = open(path.join(main_tests_dir, 'pbr_structure.inp'), 'r')
    print([x for x in file_], file=regtest)

def test_build_pbr_core_no_dimples(regtest):
    pbr_instance = pbr.PebbleBedReactor(output_dir=main_tests_dir, create_dimples=False)
    pbr_instance.build_pbr_core()
    pbr_instance.write_pbr_core()
    file_ = open(path.join(main_tests_dir, 'pbr_structure.inp'), 'r')
    print([x for x in file_], file=regtest)

def test_build_pbr_core_simple(regtest):
    pbr_instance = pbr.PebbleBedReactor(output_dir=main_tests_dir, simple_core=True)
    pbr_instance.build_pbr_core()
    pbr_instance.write_pbr_core()
    file_ = open(path.join(main_tests_dir, 'pbr_structure.inp'), 'r')
    print([x for x in file_], file=regtest)    
