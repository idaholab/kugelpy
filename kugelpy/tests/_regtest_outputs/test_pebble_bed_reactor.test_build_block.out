{1: {'surfaces': {'block': 'surf block_1_s  pad 0.0 0.0 120.0 206.6 0.0 20.0\n', 'dimples': 'surf block_1_plane plane 0.984807753012208 -0.1736481776669303\nsurf block_1_d0_s cylv 0.0 0.0 70.0 0.984807753012208 -0.1736481776669303 0.0 17.5\nsurf block_1_d1_s cylv 0.0 0.0 140.0 0.984807753012208 -0.1736481776669303 0.0 17.5\nsurf block_1_d2_s cylv 0.0 0.0 210.0 0.984807753012208 -0.1736481776669303 0.0 17.5\nsurf block_1_d3_s cylv 0.0 0.0 280.0 0.984807753012208 -0.1736481776669303 0.0 17.5\nsurf block_1_d4_s cylv 0.0 0.0 350.0 0.984807753012208 -0.1736481776669303 0.0 17.5\nsurf block_1_d5_s cylv 0.0 0.0 420.0 0.984807753012208 -0.1736481776669303 0.0 17.5\nsurf block_1_d6_s cylv 0.0 0.0 490.0 0.984807753012208 -0.1736481776669303 0.0 17.5\nsurf block_1_d7_s cylv 0.0 0.0 560.0 0.984807753012208 -0.1736481776669303 0.0 17.5\nsurf block_1_d8_s cylv 0.0 0.0 630.0 0.984807753012208 -0.1736481776669303 0.0 17.5\nsurf block_1_d9_s cylv 0.0 0.0 700.0 0.984807753012208 -0.1736481776669303 0.0 17.5\nsurf block_1_d10_s cylv 0.0 0.0 770.0 0.984807753012208 -0.1736481776669303 0.0 17.5\nsurf block_1_d11_s cylv 0.0 0.0 840.0 0.984807753012208 -0.1736481776669303 0.0 17.5\n', 'control_rods': '', 'risers': 'surf block_1_riser_s cylz 175.78818391267913 -30.99619971354706 8.5 0.0 938.8\n', 'control_rod': 'surf block_1_sr_s cylz 130.97943115062367 -23.09520762970173 6.25 918.0 1024.8\n', 'control_rod_cavity': 'surf block_1_cr_cavity_s cylz 130.97943115062367 -23.09520762970173 6.5 0.0 1024.8\n'}, 'cells': {'block': 'cell block_1_c block_1_u reflector -block_1_s  #block_1_d0_c #block_1_d1_c #block_1_d2_c #block_1_d3_c #block_1_d4_c #block_1_d5_c #block_1_d6_c #block_1_d7_c #block_1_d8_c #block_1_d9_c #block_1_d10_c #block_1_d11_c #block_1_sr_c #block_1_cr_cavity_c #block_1_riser_c', 'dimples': 'cell block_1_d0_c pebbles_u fill pebble_bed -block_1_d0_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d1_c pebbles_u fill pebble_bed -block_1_d1_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d2_c pebbles_u fill pebble_bed -block_1_d2_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d3_c pebbles_u fill pebble_bed -block_1_d3_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d4_c pebbles_u fill pebble_bed -block_1_d4_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d5_c pebbles_u fill pebble_bed -block_1_d5_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d6_c pebbles_u fill pebble_bed -block_1_d6_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d7_c pebbles_u fill pebble_bed -block_1_d7_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d8_c pebbles_u fill pebble_bed -block_1_d8_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d9_c pebbles_u fill pebble_bed -block_1_d9_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d10_c pebbles_u fill pebble_bed -block_1_d10_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d11_c pebbles_u fill pebble_bed -block_1_d11_s -outer_dimple inner_dimple block_1_plane\n', 'control_rods': '', 'risers': 'cell block_1_riser_c block_1_riser_u helium -block_1_riser_s   ', 'control_rod': 'cell block_1_sr_c block_1_sr_u safety_rod -block_1_sr_s   ', 'control_rod_cavity': 'cell block_1_cr_cavity_c block_1_cr_cavity_u helium -block_1_cr_cavity_s   #block_1_sr_c'}, 'universes': {'block': 'cell block_1_u 0 fill block_1_u -block_1_s  #block_1_d0_c #block_1_d1_c #block_1_d2_c #block_1_d3_c #block_1_d4_c #block_1_d5_c #block_1_d6_c #block_1_d7_c #block_1_d8_c #block_1_d9_c #block_1_d10_c #block_1_d11_c #block_1_sr_c #block_1_cr_cavity_c #block_1_riser_c', 'dimples': 'cell block_1_d0_u 0 fill pebbles_u -block_1_d0_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d1_u 0 fill pebbles_u -block_1_d1_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d2_u 0 fill pebbles_u -block_1_d2_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d3_u 0 fill pebbles_u -block_1_d3_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d4_u 0 fill pebbles_u -block_1_d4_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d5_u 0 fill pebbles_u -block_1_d5_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d6_u 0 fill pebbles_u -block_1_d6_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d7_u 0 fill pebbles_u -block_1_d7_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d8_u 0 fill pebbles_u -block_1_d8_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d9_u 0 fill pebbles_u -block_1_d9_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d10_u 0 fill pebbles_u -block_1_d10_s -outer_dimple inner_dimple block_1_plane\ncell block_1_d11_u 0 fill pebbles_u -block_1_d11_s -outer_dimple inner_dimple block_1_plane\n', 'control_rods': '', 'risers': 'cell block_1_riser_u 0 fill block_1_riser_u -block_1_riser_s   ', 'control': 'cell block_1_sr_u 0 fill block_1_sr_u -block_1_sr_s   ', 'control_rod_cavity': 'cell block_1_cr_cavity_u 0 fill block_1_cr_cavity_u -block_1_cr_cavity_s   #block_1_sr_c'}}}
